core/string/src/builder.rs

```diff
@@ -0,0 +1,423 @@
+use crate::{
+    alloc_overflow, tagged::Tagged, JsString, RawJsString, RefCount, TaggedLen, DATA_OFFSET,
+};
+use std::{
+    alloc::{alloc, dealloc, realloc, Layout},
+    cell::Cell,
+    marker::PhantomData,
+    mem::ManuallyDrop,
+    ops::AddAssign,
+    ptr::{self, addr_of, addr_of_mut, NonNull},
+};
+
+#[doc(hidden)]
+mod private {
+    /// Inner elements represented for `JsStringBuilder`.
+    pub trait JsStringData {}
+
+    impl JsStringData for u8 {}
+    impl JsStringData for u16 {}
+}
```

Comment:
Module `private` is not `pub` so it can only be only accessed in `builder.rs`, so user can't use it.

Reply:
Ahh, sorry, you are right!

Actually it may be a good to expose the `JsStringData` (but seal it) we can do this by extending a `Sealed` trait that is private and implemented on `u8` and `u16`. That way the user can use the trait in generic type bounds if they want.

```rust
mod private {
    pub trait Sealed {}

    impl Sealed for u8 {}
    impl Sealed for u16 {}
}

/// Inner elements represented for `JsStringBuilder`.
pub trait JsStringBuilder: private::Sealed {}
```

```rust
// In user code:
struct SomeUserGenericStruct<T: JsStringData> { // The user can use the trait
   builder: JsStringBuilder<T>,
   /// ...
}
```

Reply:
Done.